Dds 1029 trashbin endpoints

dmlond opened the pull request.

dmlond committed:
Factories modified to create :purged resources
 - file_versions
 - data_files
 - folders

Trashbin design updated

TrashbinAPISpec started

dmlond committed:
Policy specifications expect policies to implement a restore? policy
to determine when the object can be restored.

dmlond committed:
file versions in deleted files cannot be restored

dmlond committed:
TrashbinAPISpec expects TrashbinAPI to implement:
  - get /api/v1/trashbin/:object_kind/:object_id
  - put /api/v1/trashbin/:object_kind/:object_id/restore
  - put /api/v1/trashbin/:object_kind/:object_id/purge
According to api_design/DDS-829-trash_bin.md
TrashbinAPI meets spec

dmlond committed:
Merge branch 'DDS-829-trash_bin' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-1029_trashbin_endpoints

dmann commented on app/api/dds/v1/trashbin_api.rb:
```diff
@@ -0,0 +1,129 @@
+        parent_params = declared(params, {include_missing: false}, [:parent])
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        purge_object = object_kind.find_by!(id: params[:object_id], is_deleted: true)
+        unless purge_object.class.include? Restorable
```
I feel like this check should happen using `object_kind` on the line above the call to `.find_by!`. Also, the error_json formatting seems like it belongs in a separate helper method, which would help to reduce the clutter in this endpoint.

dmann commented on app/api/dds/v1/trashbin_api.rb:
```diff
@@ -0,0 +1,129 @@
+          }
+          error!(error_json, 404)
+        end
+        unless purge_object.is_purged?
```
Nothing happens if `purge_object.is_purged?` is true? Seems like an immutable validation error or something should result from this...

dmann commented on app/api/dds/v1/trashbin_api.rb:
```diff
@@ -0,0 +1,129 @@
+        end
+        unless purge_object.is_purged?
+          purge_object.is_deleted = false
+          if purge_object.is_a?(FileVersion)
```
Could this logic be stored in the model? Possibly a `#restore` method?

dmann commented on app/api/dds/v1/trashbin_api.rb:
```diff
@@ -0,0 +1,129 @@
+        authenticate!
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        purge_object = object_kind.find_by!(id: params[:object_id], is_deleted: true)
+        unless purge_object.class.include? Restorable
```
My comments for the `/restore` endpoint apply to the `/purge` endpoint as well.

dmann reviewed (changes requested):
The grape endpoints need to be DRYed out some, probably by moving logic to the model layer and possibly raising/rescuing exceptions instead of nested if statements. All the rest looks good.

dmlond committed:
changed documentation for restore

dmlond committed:
ChildMinder behavior expects subject to implement #restore(child)
ProjectSpec expects #restore to
  - raise IncompatibleParentException if project is deleted
  - raise IncompatibleParentException if child is not a Container
  - set child.parent_id to nil
  - set child.project_id to id
Project meets spec

IncompatibleParentException and TrashbinException implemented

dmlond committed:
FolderSpec expects Folder to
  - meet ChildMinder #restore expectation
     - raise TrashbinParentException if folder is deleted
     - raise IncompatibleParentException if child is not a Container
     - set child.parent_id to id
     - set child.project_id to project_id
     - set child.is_deleted = false
Folder meets spec

dmlond committed:
DataFileSpec expects DataFile to
  - meet ChildMinder #restore expectation
     - raise TrashbinParentException if is_deleted
     - raise IncompatibleParentException if child is not one of its own
       file_versions
DataFile meets spec

dmlond committed:
refactored to meet Code Review suggestions
changed response expectations for TrashbinParentException and IncompatibleParentException
expectations.

dmlond committed:
Purgable behavior expects Purgable to implement #purge
Purgable concern implements default #purge to set is_deleted and is_purged to true
FileVersionSpec expects FileVersion to override default #purge to raise UnPurgableException

DataFile, Folder, and FileVersion meet spec
UnPurgableException implemented

dmlond committed:
Refactored /api/v1/:object_kind/:object_id/purge based on PR Code Review suggestions
modified TrashbinAPISpec expectations to match UnPurgableException message expectations

dmlond commented:
that was useful. Thinking about these code review suggestions forced me to realize that I had missed the fact that containers could be restored to project roots, among some other things. Please review again :)

dmann commented on app/models/data_file.rb:
```diff
@@ -103,4 +103,11 @@ def creator
   end
+
+  def restore(child)
+    raise TrashbinParentException.new("#{kind} #{id} is deleted, and cannot restore its versions.::Restore #{kind} #{id}.") if is_deleted?
```
Some of this error message formatting could go in the exception classes. Maybe pass them `kind` and `id`, or an object?

dmlond replied:
This is nice, but I think it can wait until the hardening phase in a couple of weeks. It isnt something that prevents code from moving forward.

dmann commented on app/api/dds/v1/trashbin_api.rb:
```diff
@@ -0,0 +1,134 @@
+        object_kind = KindnessFactory.by_kind(params[:object_kind])
+        purge_object = object_kind.find_by!(id: params[:object_id], is_deleted: true)
+        raise UnRestorableException.new(purge_object.kind) unless purge_object.class.include? Restorable
+        unless purge_object.is_purged?
```
I still don't know what to think about checking `.is_purged?` when restoring and not throwing an error or something. If it's gotten this far, `purge_object` is `is_deleted: true`, so if it's also `.is_purged?`, then trying to restore it shouldn't work, right?

dmlond replied:
I prefer keeping the check for is_purged?. It is a guard clause that quickly skips further SQL calls to get a parent or perform authorize!.

dmann reviewed (commented):
Lots of good improvements. I think the exception classes could hold some formatting logic, maybe taking the calling object as a reference. Also see my comment regarding `/restore`. 👍

dmlond commented:
Merging with the intent to refactor exceptions in a future hardening phase
[DDS-1035](https://jira.oit.duke.edu/browse/DDS-1035)
